### PR TITLE
NEUT-1765 Add Sponsored Banner SwiftUI Component

### PR DESCRIPTION
**Pull request by darioroa**
This PR adds the Sponsored Banner Backpack Component

image

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] `README.md`
+ [x] Tests
+ [x] [Screenshotting code](https://github.com/Skyscanner/backpack-ios/blob/main/Example/Backpack%20Screenshot/Screenshots.swift)
+ [ ] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/main/Backpack/Backpack.h)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            if let subheadline = subheadline {
+                BPKText(subheadline, style: .caption)
+                    .foregroundColor(variant.color)
+            }
+        }
+    }
+    
+    private var iconView: some View {
+        ZStack {
+            Text("")
```
Comment (AxelCB) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:116]: Why do we need an empty text?

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            BPKIconView(.informationCircle)
+                .foregroundColor(variant.color)
+        }
+        .accessibilityAddTraits(.isButton)
+        .accessibilityValue(callToAction?.accessibilityLabel ?? "")
+        .frame(height: 22)
+        .frame(height: 22)
```
Comment (AxelCB) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:123]: Double frame modifier? 🤔

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    }
+    
+    private var iconView: some View {
+        ZStack {
+            Text("")
+            BPKIconView(.informationCircle)
+                .foregroundColor(variant.color)
+        }
+        .accessibilityAddTraits(.isButton)
+        .accessibilityValue(callToAction?.accessibilityLabel ?? "")
```
Comment (AxelCB) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:121]: Does accessibility value with an empty string make sense? Can't we use `ConditionalModifier` here?

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                .foregroundColor(.canvasContrastColor)
+        )
+        .clipShape(
+            RoundedRectangle(cornerRadius: .sm, style: .continuous)
+        )
+    }
+    
+    private var topView: some View {
+        HStack(spacing: .md) {
+            HStack {
```
Comment (AxelCB) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:80]: Why do we need 2 Hstack levels here?

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+public struct BPKSponsoredBanner: View {
+    let logo: Image?
+    let title: String?
+    let subheadline: String?
+    let callToAction: CallToAction?
+    let bodyText: String?
+    let variant: Variant
+    let backgroundColor: Color
+    
+    @State var showTerms: Bool
```
Comment (AxelCB) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:31]: Personally I find this name feeling a bit incomplete, I'd be okay with a longer more descriptive name `showTermsAndConditions` but happy to hear what others think on short vs long names here 😄

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+struct BPKSponsoredBanner_Previews: PreviewProvider {
+    static var previews: some View {
+        BPKSponsoredBanner(
+            title: "Title",
+            subheadline: "Subheading",
+            callToAction: .init(
+                text: "Sponsored",
+                accessibilityLabel: "More Information", showIcon: false),
```
Comment (AxelCB) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:163]: should showIcon be on a separate line?

**Review by AxelCB — COMMENTED** [review]
Besides the inline comments, got a couple of questions:

1. Why are you adding duplicated snapshots under BackpackTests/FailureDiffs?
2. Should we include tests/screenshots for a sponsored banner without text/title?
3. For dark mode test & examples with bodyText/ terms and conditions, I'd consider if we should wrap the component around with a view with another colour, or even if the chosen one is correct since there's no contrast with the default canvas colour

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 6 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:163]: Yes!

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 1 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:116]: This is a workaround due to the VoiceOver issues. Investigating.

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 4 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:101]: Good catch, initially it was to have separation of the info label from the rest, but perhaps it's not needed anymore 🤔

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 3 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift]: Did a similar thing, check latest PR 😄

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 4 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift]: Done, check 98da5d5e583a83f8c6ac8ddbf181a0c07f300667

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 1 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift]: It's nicely done now 👍

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 2 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift]: Sorted 👍

**Comment by darioroa** [comment]
> Besides the inline comments, got a couple of questions:
> 
> 1. Why are you adding duplicated snapshots under BackpackTests/FailureDiffs?

Was a mistake, corrected by @gert-janvercauteren (thank you!)

> 2. Should we include tests/screenshots for a sponsored banner without text/title?

Thought about it, in fact it's the "default" initialiser, but it does not make sense implementation-wise. It's just for the case where there is, for instance, only a logo on the left and a "sponsored" text on the right (for Inline plus, for instance).

> 3. For dark mode test & examples with bodyText/ terms and conditions, I'd consider if we should wrap the component around with a view with another colour, or even if the chosen one is correct since there's no contrast with the default canvas colour

Corrected, you can take a look at the updated custom token for better contrast 👍

File: Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift [hunk 7]
```diff
@@ -36,6 +36,16 @@ public struct BPKCardWrapper<HeaderContent: View, WrappedContent: CardWrappable>
...
     
+    public init(
+        backgroundColor: UIColor,
+        @ViewBuilder header: () -> HeaderContent,
+        @ViewBuilder card: () -> WrappedContent
+    ) {
+        self.backgroundColor = BPKColor.init(value: backgroundColor)
+        self.header = header()
+        self.card = card()
+    }
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift:47]: Is this change needed in this PR?

File: Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift [hunk 8]
```diff
@@ -102,6 +112,22 @@ struct BPKCardWrapper_Previews: PreviewProvider {
...
+                        .foregroundColor(.white)
+                    Spacer()
+                }
+                .foregroundColor(.white)
+                .padding(.md)
+            } card: {
+                BPKCard(elevation: .none) {
+                    Text("Card Content")
+                }
+            }
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift:129]: This probably needs to be rolled back

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    let logo: Image?
+    let title: String?
+    let subheadline: String?
+    let callToAction: CallToAction?
+    let bodyText: String?
+    let variant: Variant
+    let backgroundColor: Color
+    let customAccessibilityLabel: String?
+    
+    @State var showTerms: Bool
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:32]: This can probably be "isExpanded" as that is what it does

Comment (gert-janvercauteren) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:32]: These probably all need to be private

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        }
+        
+        var buttonStyle: BPKButton.Style {
+            (self == .onDark) ? .linkOnDark : .link
+        }
+    }
+    
+    public struct CallToAction {
+        public let text: String
+        public let accessibilityLabel: String
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:174]: Please rename this to a11y hint,
do we think we need this in general if we can 'expand' the banner?

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+struct BPKSponsoredBanner_Previews: PreviewProvider {
+    static var previews: some View {
+        BPKSponsoredBanner(
+            title: "Title",
+            subheadline: "Subheading",
+            callToAction: .init(
+                text: "Sponsored",
+                accessibilityLabel: "More Information",
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:192]: "Double tap for more information"

**Review by gert-janvercauteren — COMMENTED** [review]
Left some remarks on the accessibility / general build.

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 9 above)
Comment (frugoman) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:32]: This! ☝️

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        )
+        .if(customAccessibilityLabel != nil) { view in
+            view
+                .accessibilityElement(children: .ignore)
+                .accessibilityLabel(customAccessibilityLabel ?? "")
+        }
+    }
+    
+    private var bodyTextView: some View {
+        Group {
```
Comment (frugoman) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:79]: can you make this a `@ViewBuilder` and remove the `Group`? that way if the text does not need to be shown nothing would be added

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+            if hasBody && showTerms == true {
+                BPKText(bodyText ?? "", style: .caption)
```
Comment (frugoman) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:81]: I think you probably want to also if-let the body text as well, to not accidentally show an empty BPKText

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+                    .scaledToFit()
+                    .frame(maxHeight: 22)
+            }
+            titlesView
+            Spacer()
+            if let callToAction = callToAction {
+                BPKText(callToAction.text, style: .caption)
+                    .foregroundColor(variant.color)
+            }
+            if callToAction?.showIcon == true {
```
Comment (frugoman) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:121]: should this line be inside the if let thats right above?

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    
+    private func toggleBodyText() {
+        withAnimation(.spring()) {
+            showTerms.toggle()
+        }
+    }
+}
+
+extension BPKSponsoredBanner {
```
Comment (frugoman) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:158]: could you move this extension to a new file? nothing bad to have it here really, just to make the files smaller and easier to read later

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    public enum Variant {
+        case onLight
+        case onDark
+        
+        var color: Backpack_SwiftUI.BPKColor {
+            (self == .onDark) ? BPKColor.textOnDarkColor : BPKColor.textOnLightColor
```
Comment (frugoman) [inline Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:164]: I think you can swftify this props (if you still need them for this to compile then it's fine)
```swift
var color: BPKColor {
  self == .onDark ? .textOnDarkColor : .textOnLIghtColor
}

File: Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift (hunk 7 above)
Comment (darioroa) [reply Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift:47]: Definitely not, made it into it somehow when merging stuff 🤔 . Checking 👍

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 10 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:174]: 🤔 Good point. This is more of a product/design level req, but if there is already an `isButton` trait, it's clear it is expandable even without this.

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    public init(
+        logo: Image? = nil,
+        title: String? = nil,
+        subheadline: String? = nil,
+        callToAction: CallToAction? = nil,
+        bodyText: String? = nil,
+        variant: Variant,
+        backgroundColor: Color,
+        showTerms: Bool = false,
+        customAccessibilityLabel: String? = nil
```
[the remark this answers is not in the document]
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:43]: Completely agree, let's check with Design about which ones are possible states so we can properly prevent them.

File: Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift (hunk 14 above)
Comment (darioroa) [reply Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:121]: Good catch!

**Review by frugoman — APPROVED** [review]
Looks pretty good!! thanks for contributing it!!
